File: src/backend/wms/src/model/wms_expedicao_model.py
from src import db
from src.model.wms_ordem_separacao_det_model import WmsOrdemSeparacaoDetModel
from src.model.wms_armazenamento_model import WmsArmazenamentoModel


class WmsExpedicaoModel(db.Model):
    __tablename__ = 'wms_expedicao'

    id = db.Column(db.Integer, primary_key=True)
    quantidade = db.Column(db.Integer)
    data_saida = db.Column(db.DateTime)
    id_wms_ordem_separacao_det = db.Column(db.Integer, db.ForeignKey('wms_ordem_separacao_det.id'))
    id_wms_armazenamento = db.Column(db.Integer, db.ForeignKey('wms_armazenamento.id'))

    wms_ordem_separacao_det_model = db.relationship('WmsOrdemSeparacaoDetModel', foreign_keys=[id_wms_ordem_separacao_det])
    wms_armazenamento_model = db.relationship('WmsArmazenamentoModel', foreign_keys=[id_wms_armazenamento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_wms_ordem_separacao_det = data.get('idWmsOrdemSeparacaoDet')
        self.id_wms_armazenamento = data.get('idWmsArmazenamento')
        self.quantidade = data.get('quantidade')
        self.data_saida = data.get('dataSaida')

    def serialize(self):
        return {
            'id': self.id,
            'idWmsOrdemSeparacaoDet': self.id_wms_ordem_separacao_det,
            'idWmsArmazenamento': self.id_wms_armazenamento,
            'quantidade': self.quantidade,
            'dataSaida': self.data_saida.isoformat(),
            'wmsOrdemSeparacaoDetModel': self.wms_ordem_separacao_det_model.serialize() if self.wms_ordem_separacao_det_model else None,
            'wmsArmazenamentoModel': self.wms_armazenamento_model.serialize() if self.wms_armazenamento_model else None,
        }

File: frontend/wms/lib/app/data/model/wms_expedicao_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:wms/app/infra/infra_imports.dart';
import 'package:wms/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class WmsExpedicaoModel {
	int? id;
	int? idWmsOrdemSeparacaoDet;
	int? idWmsArmazenamento;
	int? quantidade;
	DateTime? dataSaida;
	WmsOrdemSeparacaoDetModel? wmsOrdemSeparacaoDetModel;
	WmsArmazenamentoModel? wmsArmazenamentoModel;

	WmsExpedicaoModel({
		this.id,
		this.idWmsOrdemSeparacaoDet,
		this.idWmsArmazenamento,
		this.quantidade,
		this.dataSaida,
		this.wmsOrdemSeparacaoDetModel,
		this.wmsArmazenamentoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'quantidade',
		'data_saida',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Quantidade',
		'Data Saida',
	];

	WmsExpedicaoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idWmsOrdemSeparacaoDet = jsonData['idWmsOrdemSeparacaoDet'];
		idWmsArmazenamento = jsonData['idWmsArmazenamento'];
		quantidade = jsonData['quantidade'];
		dataSaida = jsonData['dataSaida'] != null ? DateTime.tryParse(jsonData['dataSaida']) : null;
		wmsOrdemSeparacaoDetModel = jsonData['wmsOrdemSeparacaoDetModel'] == null ? WmsOrdemSeparacaoDetModel() : WmsOrdemSeparacaoDetModel.fromJson(jsonData['wmsOrdemSeparacaoDetModel']);
		wmsArmazenamentoModel = jsonData['wmsArmazenamentoModel'] == null ? WmsArmazenamentoModel() : WmsArmazenamentoModel.fromJson(jsonData['wmsArmazenamentoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idWmsOrdemSeparacaoDet'] = idWmsOrdemSeparacaoDet != 0 ? idWmsOrdemSeparacaoDet : null;
		jsonData['idWmsArmazenamento'] = idWmsArmazenamento != 0 ? idWmsArmazenamento : null;
		jsonData['quantidade'] = quantidade;
		jsonData['dataSaida'] = dataSaida != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataSaida!) : null;
		jsonData['wmsOrdemSeparacaoDetModel'] = wmsOrdemSeparacaoDetModel?.toJson;
		jsonData['wmsArmazenamentoModel'] = wmsArmazenamentoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idWmsOrdemSeparacaoDet = plutoRow.cells['idWmsOrdemSeparacaoDet']?.value;
		idWmsArmazenamento = plutoRow.cells['idWmsArmazenamento']?.value;
		quantidade = plutoRow.cells['quantidade']?.value;
		dataSaida = Util.stringToDate(plutoRow.cells['dataSaida']?.value);
		wmsOrdemSeparacaoDetModel = WmsOrdemSeparacaoDetModel();
		wmsOrdemSeparacaoDetModel?.id = plutoRow.cells['wmsOrdemSeparacaoDetModel']?.value;
		wmsArmazenamentoModel = WmsArmazenamentoModel();
		wmsArmazenamentoModel?.id = plutoRow.cells['wmsArmazenamentoModel']?.value;
	}	

	WmsExpedicaoModel clone() {
		return WmsExpedicaoModel(
			id: id,
			idWmsOrdemSeparacaoDet: idWmsOrdemSeparacaoDet,
			idWmsArmazenamento: idWmsArmazenamento,
			quantidade: quantidade,
			dataSaida: dataSaida,
		);			
	}

	
}